dev/VSIX/Extension/Cs/Dev17/WizardImplementation.cs

```diff
@@ -0,0 +1,141 @@
+using Microsoft.VisualStudio.TemplateWizard;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+
+namespace WindowsAppSDK.TemplateUtilities
+{
+	public class NuGetPackageInstaller : IWizard
+	{
+        private string _packageId;
+        private Project _project;
+        private IComponentModel _componentModel;
+        private IVsNuGetProjectUpdateEvents _nugetProjectUpdateEvents;
+
+        public void RunStarted(object automationObject, Dictionary<string, string> replacementsDictionary, WizardRunKind runKind, object[] customParams)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            _packageId = ExtractPackageId(replacementsDictionary);
+            _componentModel = (IComponentModel)ServiceProvider.GlobalProvider.GetService(typeof(SComponentModel));
+            _nugetProjectUpdateEvents = _componentModel.GetService<IVsNuGetProjectUpdateEvents>();
+            _nugetProjectUpdateEvents.SolutionRestoreFinished += OnSolutionRestoreFinished;
+        }
+        private string ExtractPackageId(Dictionary<string, string> replacementsDictionary)
+        {
+            if (replacementsDictionary.TryGetValue("$wizarddata$", out string wizardDataXml))
+            {
+                XDocument xDoc = XDocument.Parse(wizardDataXml);
+                XNamespace ns = xDoc.Root.GetDefaultNamespace();
+                string packageId = xDoc.Descendants(ns + "package")
+                                      .Attributes("id")
+                                      .Select(attr => attr.Value)
+                                      .FirstOrDefault();
+
+                if (!string.IsNullOrEmpty(packageId))
+                {
+                    return packageId;
+                }
+            }
+            return null;
+        }
+        public void ProjectFinishedGenerating(Project project)
+		{
+			_project = project;
+		}
+		public void BeforeOpeningFile(ProjectItem _)
+		{
+		}
+		public void ProjectItemFinishedGenerating(ProjectItem _)
+		{
+		}
+		public void RunFinished()
+		{
+
+		}
+		private void OnSolutionRestoreFinished(IReadOnlyList<string> projects)
+		{
+			// Debouncing prevents multiple rapid executions of 'InstallNuGetPackageAsync'
+			// during solution restore.
+			_nugetProjectUpdateEvents.SolutionRestoreFinished -= OnSolutionRestoreFinished;
+			var joinableTaskFactory = new JoinableTaskFactory(ThreadHelper.JoinableTaskContext);
+			joinableTaskFactory.RunAsync(InstallNuGetPackageAsync);
+
+		}
+		private Task InstallNuGetPackageAsync()
+		{
+            if (string.IsNullOrEmpty(_packageId))
+            {
+                string message = "Failed to install the NuGet package. The package ID provided in the template configuration is either missing or invalid. Please ensure the template is correctly configured with a valid package ID.";
```

Comment:
generally, I prefer to see diagnostics nearer the point of discovery (line 51 in this case)